Archive cache fixtures (#67)
aiidateam/aiida-testing: docs/source/user_guide/archive_cache.rst

===========================
Using :mod:`.archive_cache`
===========================

:mod:`.archive_cache` provides three central fixtures 

> - :py:func:`~aiida_testing.archive_cache.create_node_archive` and :py:func:`~aiida_testing.archive_cache.load_node_archive` are used to manage AiiDA archive test files in the pytest environment

Reviewer: Do we still need these?

Given that they are not documented below, I suggest you either move the mention of them to the end of the document or you remove them (fixtures + documentation) entirely.

It appears new users can really focus just on the `enable_archive_cache` fixture.

Author: transformed these into normal functions and moved them into the utils module 👍

    """
    Very simple workchain which wraps a diff calculation for testing purposes
    """

    @classmethod
    def define(cls, spec):
        super(DiffWorkChain, cls).define(spec)
        spec.expose_inputs(DiffCalculation, namespace='diff')
        spec.outline(
            cls.rundiff,
            cls.results,
        )
        spec.output('computed_diff')

    def rundiff(self):
        inputs = self.exposed_inputs(DiffCalculation, 'diff')
        running = self.submit(DiffCalculation, **inputs)

        return ToContext(diff_calc=running)

    def results(self):
        computed_diff = self.ctx.diff_calc.get_outgoing().get_node_by_label('diff')
        self.out('computed_diff', computed_diff)

And an example of using the above fixtures in a pytest test suite looks as follows

.. code-block:: python

    def test_diff_workchain(enable_archive_cache):
        """
        End to end test of DiffWorkchain
        """
        # ... set up test inputs

        inputs = {
            'diff': {
                'code': diff_code,
                'parameters': parameters,
                'file1': file1,
                'file2': file2,
            }
        }

>     with enable_archive_cache('diff_workchain.aiida'):

Reviewer: indent, please

        res, node = run_get_node(DiffWorkChain, **inputs)

    # Test results of workchain

The fixture will look for the AiiDA archive named ``diff_workchain.aiida`` in a folder named ``caches`` in the same directory as the test file, if nothing else is specified.
If this exists the archive is imported and the AiiDA caching functionality is enabled. All calculations created inside the with block will use the cached nodes if their
inputs and attributes match.

> If the archive does not exist the workchain will try to run the complete calculation.

Reviewer: ```suggestion
If the archive does not exist the workchain will try to run the complete calculation and, afterwards, create the archive in the specified location.
```
am I right?

Author: yes

.. note::

>     The caching mechanism of AiiDA is modified within test functions using these fixtures to ignore attributes, that would break the
>     caching if the tests are run on different machines with different versions of AiiDA for example
>     By default the stored test caches are also migrated to match the installed AiiDA version

Reviewer: ```suggestion
    The hashing mechanism of AiiDA is modified within tests that use the fixture to ignore certain attributes that would invalidate the
    cache when running the tests on different machines, with different versions of AiiDA, etc.
    By default, the test archives are migrated to match the installed AiiDA version.
```

Reviewer: I guess this means you are ignoring the Computer on which the Code is set up.
Are you including any information about the Code in the hash?

Author: Only the name of the calculation plugin is included in the hash of the code

> The following options can be specified in the ``aiida-testing-config.yml`` file

Reviewer: Maybe mention here that none of these options need to be modified/set in order to use `archive_cache` (I guess it will work just out of the box?).

.. code-block:: yaml

    archive_cache:
        default_data_dir: ... #If specified all relative paths passed to enable_archive_cache are relative to this
        ignore:

>             calcjob_inputs: [...] #List of link labels of inputs to ignore in the aiida hash

Reviewer: can you give an example in the docs of when you may need to modify this?

Author: I wrote one paragraph giving the example of testing across AiiDA 1.X to 2.X

===========================
Using :mod:`.archive_cache`
===========================

> :mod:`.archive_cache` provides three central fixture
> 
> - :py:func:`~aiida_testing.archive_cache.enable_archive_cache` uses the AiiDA caching mechanisms to enable end-to-end tests of high-level workchains without executing individual `Calcjob` processes, if they can be cached from a stored archive

Reviewer: ```suggestion
:mod:`.archive_cache` provides the :py:func:`~aiida_testing.archive_cache.enable_archive_cache` pytest fixture. It uses the AiiDA caching mechanisms to enable end-to-end tests of high-level workchains without executing individual `Calcjob` processes, if they can be cached from a stored archive.
```

> Here an example for using the :py:func:`~aiida_testing.archive_cache.enable_archive_cache` to test a simple workchain using the ``diff`` code

Reviewer: ```suggestion
Here an example of using the :py:func:`~aiida_testing.archive_cache.enable_archive_cache` fixture to test a simple workchain using the ``diff`` code
```

The workchain looks as follows

.. code-block:: python

    class DiffWorkChain(WorkChain):
    """
    Very simple workchain which wraps a diff calculation for testing purposes
    """

    @classmethod
    def define(cls, spec):
        super(DiffWorkChain, cls).define(spec)
        spec.expose_inputs(DiffCalculation, namespace='diff')
        spec.outline(
            cls.rundiff,
            cls.results,
        )
        spec.output('computed_diff')

    def rundiff(self):
        inputs = self.exposed_inputs(DiffCalculation, 'diff')
        running = self.submit(DiffCalculation, **inputs)

        return ToContext(diff_calc=running)

    def results(self):
        computed_diff = self.ctx.diff_calc.get_outgoing().get_node_by_label('diff')
        self.out('computed_diff', computed_diff)

> And an example of using the above fixtures in a pytest test suite looks as follows

Reviewer: ```suggestion
And here is the test
```

.. code-block:: python

    def test_diff_workchain(enable_archive_cache):
        """
        End to end test of DiffWorkchain
        """
        # ... set up test inputs

        inputs = {
            'diff': {
                'code': diff_code,
                'parameters': parameters,
                'file1': file1,
                'file2': file2,
            }
        }
        with enable_archive_cache('diff_workchain.aiida'):
            res, node = run_get_node(DiffWorkChain, **inputs)

        # Test results of workchain

The fixture will look for the AiiDA archive named ``diff_workchain.aiida`` in a folder named ``caches`` in the same directory as the test file, if nothing else is specified.
If this exists the archive is imported and the AiiDA caching functionality is enabled. All calculations created inside the with block will use the cached nodes if their
inputs and attributes match.

> If the archive does not exist the workchain will try to run the complete calculation and, afterwards, create the archive in the specified location.

Reviewer: ```suggestion
If the archive does not exist the, workchain will try to run the complete calculation and, afterwards, create the archive in the specified location.
```

    The hashing mechanism of AiiDA is modified within tests that use the fixture to ignore certain attributes that would invalidate the
    cache when running the tests on different machines, with different versions of AiiDA, etc.
    By default, the test archives are migrated to match the installed AiiDA version.

The following options can be specified in the ``aiida-testing-config.yml`` file. All of the below options are optional and do not need to be modified in order to use
the archive cache functionalities

.. code-block:: yaml

    archive_cache:
        default_data_dir: ... #If specified all relative paths passed to enable_archive_cache are relative to this
        ignore:
            calcjob_inputs: [...] #List of link labels of inputs to ignore in the aiida hash
            calcjob_attributes: [...] #List of attributes of CalcjobNodes to ignore in the aiida hash
            node_attributes: #mapping of entry points to list of attributes to ignore in hashing of nodes with those entry points
                diff: [..]

An example, where it might be necessary to modify the options in the ``ignore`` namespace, is testing workchains across multiple AiiDA core versions.

> When using a AiiDA archive created in version ``1.6`` and testing the workchain with this archive in version ``2.X`` calcjob nodes contain new metadata

Reviewer: ```suggestion
When using a AiiDA archive created in version ``1.6`` and testing the workchain with this archive in version ``2.X``, calcjob nodes contain new metadata
```
